### PR TITLE
Revert "Revert "feat: markdown support""

### PR DESCRIPTION
**Pull request by ifaoji**
Reverts 42-student-council/website#200

### CODE REVIEW

**Comment by itislu** [comment]
> One issue I noticed is that empty lines get completely discarded. Even multiple newlines don't create an empty line for spacing.
> This means that comments and issues that were formatted with empty lines between paragraphs now have all the same spacing between the lines.
>
> Also, paragraphs (`#`) don't seem to work.

https://github.com/42-student-council/website/pull/195#issuecomment-2395486676

**Comment by leeoocca** [comment]
> One issue I noticed is that empty lines get completely discarded. Even multiple newlines don't create an empty line for spacing.

Empty lines without content get automatically discarded, similarly to what happens on GitHub. This solves the problem of people writing just new lines in a comment.

> This means that comments and issues that were formatted with empty lines between paragraphs now have all the same spacing between the lines.

From my local tests, one line break is always preserved. So this shouldn't break (pun not intended) previous comments rendering. The only difference is that now you only have consistent line breaks whether you inserted one, two, or more new lines. On GitHub there's a difference between one and two new lines, but ~I couldn't recreate it here.~ I just managed to recreate it. I'll push it now.

All of this is not conformant with CommonMark (one new line is not enough for a new paragraph) and is implemented with a [plugin](https://github.com/remarkjs/remark-breaks).

cc @itislu

**Comment by leeoocca** [comment]
> Also, paragraphs (#) don't seem to work.

Yes, I decided not to allow them as they can be very distracting, especially in comments. Btw they're **h**eadings (`h1`, `h2`, `h3`, ...) not **p**aragraphs (`p`).

**Review by itislu — CHANGES_REQUESTED** [review]
Multi line code blocks ` ``` ` don't work
image

Single line code blocks do work across multiple lines, but they put all text into the same line.

---

As for multiple line breaks, it's up to preference ofc therefore I leave it up to the new council, but my opinion is that they do not separate the blocks of text enough.
With the smaller text size of the issue description (same as the comments), long issues are very dense to read now.

I think an empty line should at least be the same amount as the line break introduced by the beginning of a list (but I have no idea if that's possible).
image

**Comment by leeoocca** [comment]
> Multi line code blocks ``` don't work

Yes, I didn't enable it. Why would anyone discuss code on this platform? It also allows for unlimited amount of new lines, exactly what I was trying to avoid 😅 

> they do not separate the blocks of text enough

I can separate the paragraphs more via CSS but I'm not sure it's gonna lead to good-looking usage.

**Comment by leeoocca** [comment]
I think the PR might finally be ready!

image

**Comment by itislu** [comment]
Yuhu, it's really good now!

Only one thing - what do you think about the whitespace getting removed within a line?

**Original:**
image

**New:**
image

File: src/app/components/Markdown.tsx [hunk 1]
```diff
@@ -1,84 +1,96 @@
...
+            return [SKIP, index] as const;
+        });
+    };
+};
 
-                    return <H2 {...rest}>{children}</H2>;
+export default function Markdown(props: Options & { extraClassName: string }) {
+    return (
+        <ReactMarkdown
+            className={classNames('text-lg text-balance hyphens-auto break-words [&>*]:mb-3', props.extraClassName)}
```
Comment (itislu) [inline src/app/components/Markdown.tsx:57]:
```suggestion
            className={classNames('text-lg whitespace-pre-wrap text-balance hyphens-auto break-words [&>*]:mb-3', props.extraClassName)}
```

Comment (leeoocca) [reply src/app/components/Markdown.tsx:57]: Uhmm, I just checked and GitHub behaves the same way. Might be a good use case for the `pre` element... (`` ``` `` in markdown). I'll enable it.

Comment (itislu) [reply src/app/components/Markdown.tsx:57]: Aah yes, good point. That would solve it (at least for future comments).

**Review by itislu — APPROVED** [review]

**Pull request merged**